Create a GET all endpoint for survey responses in json

sssoja opened the pull request.

nandanrao commented on dashboard-server/queries/responses/response.test.js:
```diff
@@ -57,33 +65,84 @@ describe('Response queries', () => {
+        email: 'test3@vlab.com',
+        survey: 'Survey',
+      });
+      responses.length.should.equal(8);
```
Great job with other users, but also would be good to have some responses that aren't returned!

nandanrao commented on dashboard-server/api/responses/response.routes.js:
```diff
@@ -3,6 +3,7 @@ const controller = require('./response.controller');
 
 router
   .get('/', controller.getAll)
+  .get('/first-and-last', controller.getFirstAndLast)
```
Make note that this is a breaking change, so anything in the system that was using the previous get All ("/") will break! In particular the dashboard-client.

sssoja replied:
Good point. I've changed the path back to `/` from `response.routes` but have left the naming on the `controller` as `getFirstAndLast`.

nandanrao commented on dashboard-server/queries/responses/response.test.js:
```diff
@@ -57,33 +71,151 @@ describe('Response queries', () => {
+        survey: survey.survey_name,
+      });
+
+      // responses.should.equal([
```
Why commented out?

juanpc10 commented on dashboard-server/queries/responses/response.test.js:
```diff
@@ -57,33 +71,188 @@ describe('Response queries', () => {
+      });
+
+      responses.should.equal([
+        {
+          parent_surveyid: `${survey.id}`,
+          parent_shortcode: '101',
+          surveyid: `${survey.id}`,
+          flowid: '100004',
+          userid: '126',
+          question_ref: 'ref',
+          question_idx: '10',
+          question_text: 'text',
+          response: '{ "text": "first" }',
+          timestamp: `${timestamp3}`,
+          metadata: null,
+          pageid: null,
+          translated_response: null,
+        },
```
Is this the response I can base of? @sssoja

sssoja replied:
Yes, this is the structure of the JSON object returned by the `Response.all` query. 

An example response would like this:

```
[
  {
    parent_surveyid: 'be5ae9dd-0189-478e-8a3d-4d8ead8240a4',
    parent_shortcode: '101',
    surveyid: 'be5ae9dd-0189-478e-8a3d-4d8ead8240a4',
    flowid: '100004',
    userid: '126',
    question_ref: 'ref',
    question_idx: '10',
    question_text: 'text',
    response: '{ "text": "first" }',
    timestamp: '2022-06-06 09:58:00+00:00',
    metadata: null,
    pageid: null,
    translated_response: null
  },
  {
    parent_surveyid: 'be5ae9dd-0189-478e-8a3d-4d8ead8240a4',
    parent_shortcode: '101',
    surveyid: 'be5ae9dd-0189-478e-8a3d-4d8ead8240a4',
    flowid: '100001',
    userid: '126',
    question_ref: 'ref',
    question_idx: '10',
    question_text: 'text',
    response: '{ "text": "last" }',
    timestamp: '2022-06-06 10:00:00+00:00',
    metadata: null,
    pageid: null,
    translated_response: null
  }
]
```

sssoja reviewed (commented):
Hey @nandanrao , I'm a bit stuck on this. I keep getting the following error when testing the   the get all route `TypeError: Cannot read properties of undefined (reading 'query')`. 

The error traces back to line 36 and 53 of `response.queries` and line 34 of the `response.controller`.

<img width="887" alt="Screenshot 2022-06-21 at 17 00 32" src="https://user-images.githubusercontent.com/53219789/174832267-b3daa8f6-1db5-41b7-9e60-f94ded6c43d7.png">


As far as I understand, `this` is evaluating to undefined and so it can't read `query`. This only happens when I test the `all` endpoint from `getAll()`. The `_all()` function as a standalone function still returns all the responses as it should.

nandanrao commented:
Take a look at the line reference in the trace -> line 36. It's the `this.query` - remember that `this` is supposed to be an instance of the `Pool` class from node-pg, elsewhere in the code. So probably you need to make sure you have a `pool` there as well.

sssoja commented:
> Take a look at the line reference in the trace -> line 36. It's the `this.query` - remember that `this` is supposed to be an instance of the `Pool` class from node-pg, elsewhere in the code. So probably you need to make sure you have a `pool` there as well.

I updated the `all` function which is currently binded to `pool` so that it passes `this` as a param to the `_all` query. I then reference that instance as a param called `pool` in the `_all` query. I now get a different error `pool.query is not a function` rather than `this` reading as undefined. 

In trying to isolate, I logged out `pool` and I get the name of the survey `Survey123` but nothing else. I've been reading up on pooling from the node-postgres docs, specifically `pool.query` and as far as I can see the implementation is correct: 

```
const { Pool } = require('pg')
const pool = new Pool()
pool
  .query(QUERY)
```

So I'm wondering if I've got my use of `this` and `pool` muddled...

nandanrao commented on dashboard-server/queries/responses/response.queries.js:
```diff
@@ -33,8 +36,9 @@ async function _all({ email, survey, timestamp, userid, ref, pageSize }, pool) {
   ORDER BY (timestamp, responses.userid, question_ref)
   LIMIT $6`;
 
+  pool = new Pool(DATABASE_CONFIG);
```
Wait, no, this shouldn't be necessary! Look at how all the other functions in the controllers access pool. Follow the same pattern as the other developers - find the pattern!

nandanrao replied:
@sssoja just pinging you on this.

nandanrao commented on dashboard-server/queries/responses/token.js:
```diff
@@ -0,0 +1,31 @@
+  }
+
+  default() {
+    const firstToken = "1970-01-01 00:00:00+00:00/''/''";
```
I'm curious that you need the single quotes at all here...

nandanrao commented on dashboard-server/queries/responses/response.test.js:
```diff
@@ -57,34 +73,357 @@ describe('Response queries', () => {
+      });
+
+      describe('surveyNotFound', () => {
+        it('should return no responses if the survey name is not found', async () => {
```
I'm not convinced it should return an empty array if the survey isn't found, as opposed to a 404!

nandanrao commented on dashboard-server/queries/responses/response.test.js:
```diff
@@ -57,34 +74,351 @@ describe('Response queries', () => {
+          answers.length.should.equal(4);
+
+          let tokenReceived = firstResponse.body.items[0].token;
+          tokenReceived.should.equal('MTk3MC0wMS0wMSAwMDowMDowMCswMDowMA==');
```
I wouldn't even assert what the token looks like. You just need to assert that the token _works_ - that it gets you a new set of data. Does that make sense?

nandanrao commented on dashboard-server/queries/responses/response.test.js:
```diff
@@ -57,34 +73,318 @@ describe('Response queries', () => {
+        });
+      });
+
+      describe('ROUTE /all', () => {
```
Great!

nandanrao commented:
Looking good!

sssoja commented:
@nandanrao I just pushed a few more commits to reflect the changes discussed in standup. 

I added a try-catch() to the `all()` function and refactored the error message to one generic `RequestError` so as to not get distracted by missing users credentials.

I'm just stuck on an error on my tests for `surveyNotFound` and `userNotFound` where my `res` is returning undefined. Might just need to look at it with fresh eyes but let me know if you can spot where I'm going wrong.

nandanrao commented on dashboard-server/server.js:
```diff
@@ -21,7 +21,7 @@ app
   .use('/health', (req, res) => {
     // TODO: check connection to DB
-    return res.status(200).send('OK');
+    return res.status(200).send('hola');
```
???

nandanrao commented on dashboard-server/api/responses/response.controller.js:
```diff
@@ -4,23 +4,42 @@ const { Readable } = require('stream');
+exports.getAll = async (req, res) => {
+  try {
+    const { survey, after, pageSize } = req.query;
+    const { email } = req.user; // req is mutable and we can add whatever we want
```
Cleanup comment!

nandanrao commented on dashboard-server/queries/responses/token.js:
```diff
@@ -0,0 +1,18 @@
+const decode = base64.decode;
+const encode = base64.encode;
+
+class Token {
```
The way you've implemented the class, it doesn't have any internal state. 

Why not just make this a module with "encode" and "decode" ?

Then instead of: 

```
const t = require('./token');
const token = new t.Token();
```

You can just do: 

```
const token = require('./token');
```

nandanrao commented on dashboard-server/queries/responses/response.test.js:
```diff
@@ -57,34 +78,316 @@ describe('Response queries', () => {
+        });
+
+        it('should return a fulfilled promise', async () => {
+          return Promise.resolve(Response.all(email, surveyName, after)).should
```
Promise.resolve? This looks very confusing. I'm not sure why something returns a resolved promise. It seems to me that Response.all returns a promise and you wait for it to be resolved to get the value...

nandanrao commented on dashboard-server/queries/responses/response.test.js:
```diff
@@ -57,34 +78,316 @@ describe('Response queries', () => {
+          responses.map(r => r.should.have.property('token'));
+
+          // second request
+          let exampleToken = responses[0].token;
```
lets clean this variable name up to not be "exampleToken", that's confusing!

nandanrao commented on dashboard-server/queries/responses/response.test.js:
```diff
@@ -57,34 +77,316 @@ describe('Response queries', () => {
+      }')
+       ,('${survey.id}', '101', '${
+        survey.id
+      }', '101', 100004, '127', 'ref', 10, 'text', '{ "text": "first" }', '6789', '${
```
make response a simple string ("first")
